drivers/flash/bch.c

```diff
@@ -0,0 +1,446 @@
+/* 
+ * Copyright (c) 2022 Macronix International Co., Ltd.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include "bch.h"
+#include <drivers/spi.h>
+
+int fls(int x)
+{
+    int r = 32;
+
+    if (!x) {
+        return 0;
+    }
+    if (!(x & 0xffff0000u)) {
+        x <<= 16;
+        r -= 16;
+    }
+    if (!(x & 0xff000000u)) {
+        x <<= 8;
+        r -= 8;
+    }
+    if (!(x & 0xf0000000u)) {
+        x <<= 4;
+        r -= 4;
+    }
+    if (!(x & 0xc0000000u)) {
+        x <<= 2;
+        r -= 2;
+    }
+    if (!(x & 0x80000000u)) {
+        x <<= 1;
+        r -= 1;
+    }
+    return r;
+}
+
+void bch_encode(struct bch_code *bch, unsigned char *data, unsigned int *ecc)
```

Comment:
Shouldn't there be some unit testing provided for the ecc functions?

Reply:
These functions have been tested and passed. When using SoftwareECC to read data, it will be called.